nilaway panic: interface conversion / send on closed channel

Here's the full trace:

```bash
$ nilaway ./...
panic: interface conversion: types.Object is *types.Var, not *types.Func [recovered]
	panic: send on closed channel

goroutine 172328 [running]:
go.uber.org/nilaway/assertion/function.analyzeFunc.func1()
	/home/user/.gvm/pkgsets/go1.20/global/pkg/mod/go.uber.org/nilaway@v0.0.0-20231031155528-970a3b8379e3/assertion/function/analyzer.go:457 +0x11e
panic({0x7972e0, 0xc1a7b6e210})
	/home/user/.gvm/gos/go1.20/src/runtime/panic.go:884 +0x213
go.uber.org/nilaway/assertion/function/assertiontree.exprAsAssignmentConsumer.func3({0x878f68, 0xc071ea71c0})
	/home/user/.gvm/pkgsets/go1.20/global/pkg/mod/go.uber.org/nilaway@v0.0.0-20231031155528-970a3b8379e3/assertion/function/assertiontree/backprop_util.go:487 +0x955
go.uber.org/nilaway/assertion/function/assertiontree.exprAsAssignmentConsumer(0xc0e871b500, {0x878650?, 0xc071ea4900}, {0x0?, 0x0})
	/home/user/.gvm/pkgsets/go1.20/global/pkg/mod/go.uber.org/nilaway@v0.0.0-20231031155528-970a3b8379e3/assertion/function/assertiontree/backprop_util.go:593 +0x679
go.uber.org/nilaway/assertion/function/assertiontree.backpropAcrossSend(0x878650?, 0xc071ea4900)
	/home/user/.gvm/pkgsets/go1.20/global/pkg/mod/go.uber.org/nilaway@v0.0.0-20231031155528-970a3b8379e3/assertion/function/assertiontree/backprop.go:126 +0xb3
go.uber.org/nilaway/assertion/function/assertiontree.backpropAcrossNode(0x0?, {0x878650?, 0xc071ea4900?})
	/home/user/.gvm/pkgsets/go1.20/global/pkg/mod/go.uber.org/nilaway@v0.0.0-20231031155528-970a3b8379e3/assertion/function/assertiontree/backprop.go:76 +0x12f
go.uber.org/nilaway/assertion/function/assertiontree.backpropAcrossBlock(0xc0e871b500, 0xc0e6ae8d80)
	/home/user/.gvm/pkgsets/go1.20/global/pkg/mod/go.uber.org/nilaway@v0.0.0-20231031155528-970a3b8379e3/assertion/function/assertiontree/backprop.go:42 +0x9a
go.uber.org/nilaway/assertion/function/assertiontree.BackpropAcrossFunc({0x879e20, 0xc19991d410}, 0x466fee?, 0xc071ea49c0, {0xc071ea49c0, 0x0, 0xc07f7b08f0, 0xc19991d710, 0xc19991d6e0, 0xc19991d3e0, ...}, ...)
	/home/user/.gvm/pkgsets/go1.20/global/pkg/mod/go.uber.org/nilaway@v0.0.0-20231031155528-970a3b8379e3/assertion/function/assertiontree/backprop.go:857 +0x8de
go.uber.org/nilaway/assertion/function.analyzeFunc({0x879e20?, 0xc19991d410?}, 0xc07f7b08f0, 0xc071ea49c0, {0xc071ea49c0, 0x0, 0xc07f7b08f0, 0xc19991d710, 0xc19991d6e0, 0xc19991d3e0, ...}, ...)
	/home/user/.gvm/pkgsets/go1.20/global/pkg/mod/go.uber.org/nilaway@v0.0.0-20231031155528-970a3b8379e3/assertion/function/analyzer.go:463 +0x11e
created by go.uber.org/nilaway/assertion/function.run
	/home/user/.gvm/pkgsets/go1.20/global/pkg/mod/go.uber.org/nilaway@v0.0.0-20231031155528-970a3b8379e3/assertion/function/analyzer.go:208 +0x11ca
```

--- a/assertion/function/analyzer.go
+++ b/assertion/function/analyzer.go
@@ -450,6 +450,13 @@ func analyzeFunc(
 	funcChan chan functionResult,
 	wg *sync.WaitGroup,
 ) {
+	// Deferred statements are pushed to a stack, which are executed in LIFO order. Calling
+	// wg.Done() would signal the main process that this goroutine is done, and the main process
+	// will close the result channel. However, our panic recovery handler still needs access to
+	// the result channel to send the error back. Therefore, we _must_ call `wg.Done()` after the
+	// panic recovery handler (meaning we defer it first).
+	defer wg.Done()
+
 	// As a last resort, convert the panics into errors and return.
 	defer func() {
 		if r := recover(); r != nil {
@@ -457,7 +464,6 @@ func analyzeFunc(
 			funcChan <- functionResult{err: e, index: index, funcDecl: funcDecl}
 		}
 	}()
-	defer wg.Done()
 
 	// Do the actual backpropagation.
 	funcTriggers, err := assertiontree.BackpropAcrossFunc(ctx, pass, funcDecl, funcContext, graph)

--- a/assertion/function/analyzer_test.go
+++ b/assertion/function/analyzer_test.go
@@ -28,8 +28,10 @@ import (
 	"go.uber.org/nilaway/assertion/function/assertiontree"
 	"go.uber.org/nilaway/assertion/function/functioncontracts"
 	"go.uber.org/nilaway/config"
+	"golang.org/x/tools/go/analysis"
 	"golang.org/x/tools/go/analysis/analysistest"
 	"golang.org/x/tools/go/analysis/passes/ctrlflow"
+	"golang.org/x/tools/go/cfg"
 )
 
 func TestTimeout(t *testing.T) {
@@ -104,16 +106,16 @@ func TestAnalyzeFuncPanic(t *testing.T) {
 
 	resultChan := make(chan functionResult)
 	var wg sync.WaitGroup
-	funcContext := assertiontree.FunctionContext{}
+	wg.Add(1)
+
 	// Intentionally give bad input data to cause a panic. We should convert the panic to an error
 	// and send it back to the original channel.
-	wg.Add(1)
 	go analyzeFunc(ctx,
-		nil,         /* pass */
-		nil,         /* funcDecl */
-		funcContext, /* funcContext */
-		nil,         /* graph */
-		0,           /* index */
+		&analysis.Pass{},                /* pass */
+		&ast.FuncDecl{},                 /* funcDecl */
+		assertiontree.FunctionContext{}, /* funcContext */
+		&cfg.CFG{},                      /* graph */
+		0,                               /* index */
 		resultChan,
 		&wg,
 	)

--- a/assertion/function/analyzer_test.go
+++ b/assertion/function/analyzer_test.go
@@ -97,6 +97,41 @@ func TestTimeout(t *testing.T) {
 	}
 }
 
+func TestAnalyzeFuncPanic(t *testing.T) {
+	t.Parallel()
+
+	ctx := context.Background()
+
+	resultChan := make(chan functionResult)
+	var wg sync.WaitGroup
+	funcContext := assertiontree.FunctionContext{}
+	// Intentionally give bad input data to cause a panic. We should convert the panic to an error
+	// and send it back to the original channel.
+	wg.Add(1)
+	go analyzeFunc(ctx,
+		nil,         /* pass */
+		nil,         /* funcDecl */
+		funcContext, /* funcContext */
+		nil,         /* graph */
+		0,           /* index */
+		resultChan,
+		&wg,
+	)
+	// Fire up another goroutine that waits for the work to be done and closes the result channel.
+	go func() {
+		wg.Wait()
+		close(resultChan)
+	}()
+
+	select {
+	case res := <-resultChan:
+		require.Equal(t, res.index, 0)
+		require.ErrorContains(t, res.err, "panic")
+	case <-time.After(10 * time.Second):
+		require.Fail(t, "analyzeFun did not return within 10 seconds.")
+	}
+}
+
 func TestMain(m *testing.M) {
 	goleak.VerifyTestMain(m)
 }

